setup.py

```diff
@@ -7,7 +8,23 @@ def get_install_requires():
     with open('requirements.txt', 'r') as requirements_file:
         # TODO: respect hashes in requirements.txt file
         res = requirements_file.readlines()
-        return [req.split(' ', maxsplit=1)[0] for req in res if req]
+
+        requirements = []
+        for req in res:
+            req = req.strip('\n')
+            if not req:
+                continue
+
+            if req.startswith('-e'):  # replace with PEP-508 specification
+                url = req.strip('-e ')
+                pkg = url.split('#egg=', maxsplit=1)[-1]
+                dep = pkg + " @ " + url
+            else:
+                dep = req.split(' ', maxsplit=1)[0]
+
+            requirements.append(dep)
+
+    return requirements
```

Comment:
A proper solution for this is to use `dependency_links`. 

Why there is a need to use igitt from the git repo? Is IGitt depending on urllib3?

Reply:
Why do you think that `dependency_links` is the proper solution? I am not so sure about that, especially when using pip which may run into trouble or explicitly needs `--process-dependency-links` argument (for sure in earlier versions, not 100% sure about the newer ones). There's plenty of discussions around that and I suppose it boils down to the [PEP 508] https://www.python.org/dev/peps/pep-0508/ being the accepted standard for dependency specification. You might wanna check the SO as well (for example [this](https://stackoverflow.com/questions/12518499/pip-ignores-dependency-links-in-setup-py)).
That's why I chose this approach.

As far as the igitt dependency is concerned, I am merely mimicking the entry from Pipfile to preserve consistency, that's why I used the link.

Reply:
> Why do you think that `dependency_links` is the proper solution? I am not so sure about that, especially when using pip which may run into trouble or explicitly needs `--process-dependency-links` argument (for sure in earlier versions, not 100% sure about the newer ones). There's plenty of discussions around that and I suppose it boils down to the [PEP 508](https://www.python.org/dev/peps/pep-0508/) being the accepted standard for dependency specification. You might wanna check the SO as well, for example [this](https://stackoverflow.com/questions/12518499/pip-ignores-dependency-links-in-setup-py).
> That's why I chose this approach.
> 
> As far as the igitt dependency is concerned, I am merely mimicking the entry from Pipfile to preserve consistency, that's why I used the link.

You declared in requirements.txt that you install package requirements from a git repo, but then, in setup py, you basically provide igitt directly by parsing the URL. Why would you do that?

Reply:
I am not sure I understand ... I just changed the format so that it is compatible with PEP 508, it still uses the repo.

Reply:
It uses the repo when Pipenv is used (as the repo is stated in Pipfile/Pipfile.lock), but the requirement specified for kebechet package is igitt (what you parsed from requirements.txt file) and the latest version is picked during installation based on requirements provided. The same thing was done before - that's why I don't see why you have made this change and how it fixes the version conflict.

Reply:
So first of all, the version conflict was fixed by pinning down the `requests` to the version which uses a compatible version of urllib3

And I am still confused about what you mean as far as the igitt is concerned, I might be missing something here.
Before, we used `
IGitt==0.4.1.dev20180722120744` in the requirements.txt file which is not what Pipfile specifies (which installs it from the repo), therefore I changed it in the requirements.txt to install it from the repo as well. However, `install_requires` does not understand the `-e ...` form, it understand the `<package> @ <url>` form according to the PEP specification I mentioned, which is what the few lines in the `get_install_requires` transforms it to.

Reply:
> So first of all, the version conflict was fixed by pinning down the `requests` to the version which uses a compatible version of urllib3

OK, that explains why the igitt change is not related to version conflict.

> And I am still confused about what you mean as far as the igitt is concerned, I might be missing something here.
> Before, we used ` IGitt==0.4.1.dev20180722120744` in the requirements.txt file which is not what Pipfile specifies (which installs it from the repo), therefore I changed it in the requirements.txt to install it from the repo as well. However, `install_requires` does not understand the `-e ...` form, it understands the `<package> @ <url>` form according to the PEP specification I mentioned, which is what the few lines in the `get_install_requires` transform it to.

This links back to issue https://github.com/thoth-station/kebechet/issues/103 which introduced igitt directly from master. I tried to ping igitt upstream [here](https://gitlab.com/gitmate/open-source/IGitt/issues/144) to release a new version as we rely on not yet released logic based on the issue. Having requirements pointing to a git repo is not good - we cannot manage requirements using kebechet.

As upstream is not responsive for more than year by now, it might be good idea to drop igitt and substitute it with something which has active community and development ([ogr](https://github.com/packit-service/ogr) might be a suitable option here)

BTW igitt seems to have broken versioning scheme (see the linked issue) - might be a good idea to test with "dev" releases as they are more recent as the actual releases if we can bump igitt version there for now.

Reply:
Yep, I am up for replacing igitt completely and replace it with ogr. This would fix the remaining version conflicts as well I suppose (which are in the setuptools IIRC, which igitt pins down)

Reply:
I think we should create a separate issue for that, tho, WDYT?